crates/flipperzero/src/furi/thread.rs

```diff
@@ -13,3 +136,132 @@ pub fn sleep(duration: core::time::Duration) {
         }
     }
 }
+
+/// A unique identifier for a running thread.
+#[cfg(feature = "alloc")]
+pub struct ThreadId(sys::FuriThreadId);
+
+/// A handle to a thread.
+#[cfg(feature = "alloc")]
+pub struct Thread {
+    /// Guaranteed to be UTF-8.
+    name: Option<CString>,
+    thread: *mut sys::FuriThread,
+}
+
+#[cfg(feature = "alloc")]
+impl Drop for Thread {
+    fn drop(&mut self) {
+        // TODO: is this safe to do while the thread is running?
```

Comment:
Right; I wasn't sure what managed the memory of the thread handles, but this combined with the crash from freeing the result of `sys::furi_thread_get_current()` suggests that the handles are all just pointers to the same memory. So at a minimum we need to wrap it in an `Arc` internally on the Rust side.

We also need to handle the fact that the SDK appears to expect us to clean up the thread memory that we allocate ourselves, but needs us to not do so until the firmware has finished using it.

Reply:
That's what it looks like. `sys::furi_thread_get_current()` just pulls a copy of the original `FuriThread*` that gets stashed in Thread Local Storage when the thread starts. Even the [main thread is initialised in this way](https://sourcegraph.com/github.com/flipperdevices/flipperzero-firmware@5d4057f7228b528896578ab6cf9d9a61b80a7c58/-/blob/firmware/targets/f7/src/main.c?L29:5).

This present quite a challenge, since you can't call `sys::furi_thread_free` until the thread is stopped, and the only way to guarantee that is by calling `sys::furi_thread_join`.

At present, the only way I can see of doing this safely and without potentially leaking memory is by using some sort of [scoped threads](https://doc.rust-lang.org/std/thread/fn.scope.html) that ensures that all threads spawned under the scope are joined (and `free`'d) before exiting the scope.

Since everything is in the same flat linear memory space, Furi can't even stop/cleanup all the threads after our app "exits". Pretty much the only option is to call `sys::furi_crash`.

Reply:
I've resolved the primary issue with the introduction of `FreeMan`, so we now never attempt to free a `*mut sys::FuriThread` we are holding while the thread is running.